src/super_gradients/training/transforms/keypoints/keypoints_random_affine.py

```diff
@@ -0,0 +1,193 @@
+
+            new_image_shape = sample.image.shape
+            sample.joints = self.apply_to_keypoints(sample.joints, mat_output, new_image_shape)
+
+            if sample.bboxes is not None:
+                sample.bboxes = self.apply_to_bboxes(sample.bboxes, mat_output)
+
+            if sample.areas is not None:
+                sample.areas = self.apply_to_areas(sample.areas, mat_output)
+
+            sample = sample.sanitize_sample()
+
+        return sample
+
+    @classmethod
+    def apply_to_areas(cls, areas: np.ndarray, mat):
+        det = np.linalg.det(mat[:2, :2])
+        return (areas * abs(det)).astype(areas.dtype)
+
+    @classmethod
+    def apply_to_bboxes(cls, bboxes_xywh: np.ndarray, mat: np.ndarray):
+        """
+
+        :param bboxes: (N,4) array of bboxes in XYWH format
+        :param mat:
+        :return:
+        """
+
+        def bbox_shift_scale_rotate(bbox, m):
+            x_min, y_min, x_max, y_max = bbox[:4]
+
+            x = np.array([x_min, x_max, x_max, x_min])
+            y = np.array([y_min, y_min, y_max, y_max])
+            ones = np.ones(shape=(len(x)))
+            points_ones = np.vstack([x, y, ones]).transpose()
+
+            tr_points = m.dot(points_ones.T).T
+
+            x_min, x_max = min(tr_points[:, 0]), max(tr_points[:, 0])
+            y_min, y_max = min(tr_points[:, 1]), max(tr_points[:, 1])
+
+            return np.array([x_min, y_min, x_max, y_max])
+
+        if len(bboxes_xywh) == 0:
+            return bboxes_xywh
+        bboxes_xyxy = xywh_to_xyxy(bboxes_xywh, image_shape=None)
+        bboxes_xyxy = np.array([bbox_shift_scale_rotate(box, mat) for box in bboxes_xyxy])
+        return xyxy_to_xywh(bboxes_xyxy, image_shape=None).astype(bboxes_xywh.dtype)
+
+    @classmethod
+    def apply_to_keypoints(cls, keypoints: np.ndarray, mat: np.ndarray, image_shape):
+        keypoints_with_visibility = keypoints.copy()
+        keypoints = keypoints_with_visibility[:, :, 0:2]
+
+        shape = keypoints.shape
+        dtype = keypoints.dtype
+        keypoints = keypoints.reshape(-1, 2)
+        keypoints = np.dot(np.concatenate((keypoints, keypoints[:, 0:1] * 0 + 1), axis=1), mat.T).reshape(shape)
+
+        # Update visibility status of joints that were moved outside visible area
+        image_height, image_width = image_shape[:2]
+        outside_left = keypoints[:, :, 0] < 0
+        outside_top = keypoints[:, :, 1] < 0
+        outside_right = keypoints[:, :, 0] >= image_width
+        outside_bottom = keypoints[:, :, 1] >= image_height
+
+        joints_outside_image = outside_left | outside_top | outside_right | outside_bottom
+
+        keypoints_with_visibility[:, :, 0:2] = keypoints
+        keypoints_with_visibility[joints_outside_image, 2] = 0
+        return keypoints_with_visibility.astype(dtype, copy=False)
+
+    @classmethod
+    def apply_to_image(cls, image, mat, interpolation, padding_value, padding_mode=cv2.BORDER_CONSTANT):
```

Comment:
typing

Reply:
Fixed in 88cef73f0a3c26f9e158589c0d36ef541787eaee